sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -0,0 +1,378 @@
+    INT64,   // eight-byte signed integer.
+    DECIMAL,  // Decimal integer
+    FLOAT,
+    DOUBLE,
+    STRING,  // String.
+    DATETIME, // Date and time.
+    BOOLEAN,  // Boolean.
+    ARRAY,
+    ROW;    // The field is itself a nested row.
+
+    public static final List<TypeName> NUMERIC_TYPES = ImmutableList.of(
+        BYTE, INT16, INT32, INT64, DECIMAL, FLOAT, DOUBLE);
+    public static final List<TypeName> STRING_TYPES = ImmutableList.of(STRING);
+    public static final List<TypeName> DATE_TYPES = ImmutableList.of(DATETIME);
+    public static final List<TypeName> CONTAINER_TYPES = ImmutableList.of(ARRAY);
+    public static final List<TypeName> COMPOSITE_TYPES = ImmutableList.of(ROW);
+
+    public boolean isNumericType() {
+      return NUMERIC_TYPES.contains(this);
+    }
+    public boolean isStringType() {
+      return STRING_TYPES.contains(this);
+    }
+    public boolean isDateType() {
+      return DATE_TYPES.contains(this);
+    }
+    public boolean isContainerType() {
+      return CONTAINER_TYPES.contains(this);
+    }
+    public boolean isCompositeType() {
+      return COMPOSITE_TYPES.contains(this);
+    }
+
+    /** Returns a {@link FieldTypeDescriptor} representing this primitive type. */
+    public FieldTypeDescriptor typeDescriptor() {
+      return FieldTypeDescriptor.of(this);
+    }
+  }
+
+  /**
+   * A descriptor of a single field type. This is a recursive descriptor, as nested types are
+   * allowed.
+   */
+  @AutoValue
+  public abstract static class FieldTypeDescriptor implements Serializable {
+    // Returns the type of this field.
+    public abstract TypeName getType();
+    // For container types (e.g. ARRAY), returns the type of the contained element.
+    @Nullable public abstract FieldTypeDescriptor getComponentType();
+    // For ROW types, returns the schema for the row.
+    @Nullable public abstract Schema getRowSchema();
+    /**
+     * Returns optional extra metadata.
+     */
+    @Nullable public abstract byte[] getMetadata();
+    abstract FieldTypeDescriptor.Builder toBuilder();
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setType(TypeName typeName);
+      abstract Builder setComponentType(@Nullable FieldTypeDescriptor componentType);
+      abstract Builder setRowSchema(@Nullable Schema rowSchema);
+      abstract Builder setMetadata(@Nullable byte[] metadata);
+      abstract FieldTypeDescriptor build();
+    }
+
+    /**
+     * Create a {@link FieldTypeDescriptor} for the given type.
+     */
+    public static FieldTypeDescriptor of(TypeName typeName) {
+      return new AutoValue_Schema_FieldTypeDescriptor.Builder().setType(typeName).build();
+    }
+
+    /**
+     * For container types, adds the type of the component element.
+     */
+    public FieldTypeDescriptor withComponentType(@Nullable FieldTypeDescriptor componentType) {
+      if (componentType != null) {
+        checkArgument(getType().isContainerType());
```

Comment:
i think that validations should happen in the `Builder.build()`

Reply:
so right now Builder.build() is auto generated by AutoValue.